apps/p42_log_plugin_es/src/p42_es_logger.erl

```diff
@@ -32,44 +41,73 @@ start_link(_Args) ->
 
 init([]) ->
-  {ok, Url} = application:get_env(?APP, es_url),
-  N = application:get_env(?APP, es_bulk_size, 1000),
-  {ok, _TRef} = timer:send_interval(timer:minutes(1), sync),
-  {ok, {[], Url, N}}.
-
-terminate(Reason, {Q, _Url, _N}) ->
-  io:format("PID ~p terminating for reason ~p. Queue len: ~p~n",
-            [self(), Reason, length(Q)]),
-  ok.
-
-handle_info(sync, {Q, Url, N} = State) ->
-  case length(Q) of
-    L when L == 0 -> {noreply, State};
-    _ -> send_req(Url, Q), {noreply, {[], Url, N}}
-  end;
-
-handle_info(Info, State) ->
-  io:format("Unhandled info: ~p~n", [Info]),
-  {noreply, State}.
-
-handle_cast({log, Doc}, {Q, Url, N}) ->
-  Preamble = <<"{\"index\": {}}\n">>,
-  Entry = [Preamble, jsx:encode(Doc), <<"\n">>],
-  Q1 = [Entry | Q],
-  Q2 = case length(Q1) of
-         L when L >= N -> send_req(Url, Q1), [];
-         _ -> Q1
-       end,
-  poolboy:checkin(?POOL, self()),
-  {noreply, {Q2, Url, N}}.
-
-format_status(_Opt, [_Pdict, {Q, _Url, _N}]) ->
-  [{data, [{"Queue size", length(Q)}] }].
+   {ok, _TRef} = timer:send_interval(timer:minutes(1), sync),
+   {ok,
+      #state{
+         url   = config_elastic_url()
+      ,  batch = config_elastic_batch()
+      ,  q     = q:new()
+      }
+   }.
+
+terminate(Reason, #state{q = Q}) ->
+   io:format("PID ~p terminating for reason ~p. Queue len: ~p~n",
+            [self(), Reason, q:length(Q)]),
+   ok.
+
+handle_info(sync, #state{url = Url, q = Q} = State) ->
+   case q:length(Q) of
+      L when L == 0 ->
+         {noreply, State};
+      _ ->
+         send_req(Url, Q), 
+         {noreply, State#state{q = q:new()}}
+   end;
+
+handle_info(Info, #state{} = State) ->
+   io:format("Unhandled info: ~p~n", [Info]),
+   {noreply, State}.
+
+handle_cast({log, Doc}, #state{url = Url, batch = N, q = Queue0} = State) ->
+   Queue1 = enq(Doc, Queue0),
+   Queue2 = case q:length(Queue1) of
+      L when L >= N ->
+         send_req(Url, Queue1),
+         q:new();
+      _ ->
+         Queue1
+   end,
+   poolboy:checkin(?POOL, self()),
+   {noreply, State#state{q = Queue2}}.
+
+format_status(_Opt, [_Pdict, #state{q = Q}]) ->
+  [{data, [{"Queue size", q:length(Q)}] }].
 
 %%%-------------------------------------------------------------------
 %% Internal helpers
 %%%-------------------------------------------------------------------
+enq(Doc, Queue) ->
+   Preamble = <<"{\"index\": {}}\n">>,
+   Entry = [Preamble, jsx:encode(Doc), <<"\n">>],
+   q:enq(Entry, Queue).
+
 send_req(Url, Q) ->
-  Headers = [{<<"Content-Type">>, <<"application/x-ndjson">>}],
-  Body = lists:reverse(Q),
-  hackney:request(post, Url, Headers, Body, [with_body]).
+   Headers = [{<<"Content-Type">>, <<"application/x-ndjson">>}],
+   Body = q:list(Q),
+   hackney:request(post, Url, Headers, Body, [with_body]).
+
+%%
+config_elastic_url() ->
+   os:getenv(
+      "CONFIG_ELASTIC_URL",
+      application:get_env(?APP, es_url, "http://localhost:9200/p42logs/_bulk")
```

Comment:
Default env is already set in mix.exs, which will be written to `.app` on compilation and picked up from there. Duplication here is not necessary IMO.

The OTP application environment should always reflect the values in effect in my opinion. I'm not comfortable with silently overriding it using environment variables.

Using environment variables to override app environment should be the responsibility of a separate application, or atleast a separate transient process - which will read environment on start and use that to set otp app env.

Reply:
This is very nice! I've stumbled with a deployment practices. They are not documented anywhere. You considerations is absolutely valid if proxy is distributed as a tarball to bear VMs but what about configurability of docker-based deployment. The usage of app/sys.config breaks 12-factory principles. I clearly see needs to alter config using environment variables without a docker rebuild process. Thus, I would strongly advertise usage of env variables.

I do not see any clear out-of-the box solutions that helps to mix sys.config and env variables. I've been crafting my [utility](https://github.com/fogfish/feta/blob/master/src/opts.erl) to make this. 

```erlang
// sys.config

{myapp, [
   {myconfig, {env, "CONFIG_APP", "default-value"}},
]}
...
```

```erlang
config() ->
   opts:val(myconfig, myapp).
```

I can definitely incorporate this if both ENV and sys.config support is required.

Reply:
Environment variables can only do so much for configuration, even docker supports secrets which can be arbitrary files.

Typical solutions that support system env vars fall into three categories:

1. A startup script that takes env variables, writes out a sys.config and starts erlang vm with it
2. An otp app that reads system env variables and updates otp app env
3. Each app handles reading system environment and using values from there to override defaults.

Emqx uses method 1. I would prefer method 2 or 3.

Here are my opinions:
1. Default values should not be in code. env/0 in mix.exs, or in rebar projects, .app.src is the proper place for them.
2. If I inspect otp application environment, it should truthfully reflect values in effect.
3. Any solution supporting system env vars should merely update otp app env before the values are accessed.


Since erlang term format is not a very accessible configuration format, we may have to support an alternate format, like riak, emqx, and many others. Cuttlefish and conform are two options for that. Both of those allow overriding values with system environment by writing appropriate translation functions. We may end up using those.

Opinions/thoughts?

Reply:
Based on my previous experience of development and operating Erlang apps
*  a startup script is not good. let's discard it (option 1).
* overwrites of app env at start-up or another phase is error prone. I would not recommend to procedure with option 2.
* option 3 is more handy if it is backed with proxy interface that merge sys.config with env status.
* I would not recommend to use anything else then `sys.config`